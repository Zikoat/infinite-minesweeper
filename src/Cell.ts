import Field from "./Field";

/**
 * Created by sisc0606 on 19.08.2017.
 */
export default class Cell {
  x: number;
  y: number;
  parent: Field | undefined;
  isOpen: boolean;
  isMine?: boolean;
  isFlagged: boolean;
  // the cell class only holds the information of the cell, and also mirrors
  // the functions in the field that take x/y arguments. this allows chaining,
  // like this:
  // f.getCell(10,10).open()
  constructor(
    x: number,
    y: number,
    parent: Field | undefined,
    isFlagged?: boolean,
    isMine?: boolean
  ) {
    this.x = x;
    this.y = y;
    this.parent = parent;
    this.isOpen = false;
    this.isMine = isMine;
    this.isFlagged = isFlagged;
  }

  open() {
    // opens this cell
    if (this.parent === undefined)
      console.error("i don't know my parents", this);
    return this.parent.open(this.x, this.y);
  }

  flag() {
    // flags this cell
    if (this.parent === undefined)
      console.error("i don't know my parents", this);
    return this.parent.flag(this.x, this.y);
  }
  getNeighbors() {
    //returns an array of this cell's neighbors
    if (this.parent === undefined)
      console.error("i don't know my parents", this);
    return this.parent.getNeighbors(this.x, this.y);
  }
  value() {
    if (this.parent === undefined)
      console.error("i don't know my parents", this);
    return this.parent.value(this.x, this.y);
  }
  toJSON() {
    const open = this.isOpen ? "1" : "0";
    const mine = this.isMine === undefined ? "2" : this.isMine ? "1" : "0";
    const flagged = this.isFlagged ? "1" : "0";
    return open + mine + flagged;
  }
}
