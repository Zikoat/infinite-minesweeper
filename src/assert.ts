export function assert(value: unknown, message?: string): asserts value {
  if (!value) {
    throw new Error(
      message ||
        "assertion failed, value is " + value + ", but should be truthy",
    );
  }
}
