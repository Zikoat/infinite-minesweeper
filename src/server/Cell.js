/**
 * Created by sisc0606 on 19.08.2017.
 */
export default class Cell {
	// the cell class only holds the information of the cell, and also mirrors
	// the functions in the field that take x/y arguments. this allows chaining,
	// like this:
	// f.getCell(10,10).open()
	constructor(x, y, parent=undefined, isFlagged=false, isMine=undefined){
		this.x = x;
		this.y = y;
		this.parent = parent;
		this.isOpen = false;
		this.isMine = isMine;
		this.isFlagged = isFlagged;
	}
	
	open(){// opens this cell
		if(this.parent===undefined) console.error("i don't know my parents", this);
		return this.parent.open(this.x, this.y);
	}
	
	flag(){// flags this cell
		if(this.parent===undefined) console.error("i don't know my parents", this);
		return this.parent.flag(this.x, this.y);
	}
	getNeighbors(){//returns an array of this cell's neighbors
		if(this.parent===undefined) console.error("i don't know my parents", this);
		return this.parent.getNeighbors(this.x, this.y);
	}
	value(){
		if(this.parent===undefined) console.error("i don't know my parents", this);
		return this.parent.value(this.x, this.y);
	}
	toJSON(){
		const open = this.isOpen ? "1":"0";
		const mine = this.isMine===undefined?"2":this.isMine ? "1":"0";
		const flagged = this.isFlagged ? "1":"0";
		return open + mine + flagged;
	}
}