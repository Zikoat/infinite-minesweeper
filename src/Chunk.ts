import Cell from "./Cell";

export const CHUNK_SIZE = 32;
export class Chunk{
    constructor(x,y,parent){
        this.x = x;
        this.y = y;
        this.cells = [];
        for(let i = 0; i<CHUNK_SIZE;i++){
            this.cells[i] = [];
            for(let j = 0; j<CHUNK_SIZE;j++){
                this.cells[i][j] = new Cell(x*CHUNK_SIZE+i, y*CHUNK_SIZE+j,parent);
            }
        }
    }
    getCell(x,y){
        return this.cells[x][y];
    }
    getCellFromGlobalCoords(x,y){
        return this.cells[x-this.x*CHUNK_SIZE][y-this.y*CHUNK_SIZE];
    }
    getAll(){
        let ans = [];
        var i,j;
        for(i in this.cells){
            for(j in this.cells[i]){
                ans.push(this.cells[i][j])
            }
        }
        return ans;
    }
    toJSON(){
        let ans = "";
        for(let x=0;x<CHUNK_SIZE;x++){
            for(let y=0;y<CHUNK_SIZE;y++){
                ans += this.cells[x][y].toJSON();
            }
        }
        return ans;
    }
}