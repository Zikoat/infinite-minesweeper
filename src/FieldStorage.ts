import Field from "./Field";
import Cell from "./Cell";
import { Chunk, CHUNK_SIZE } from "./Chunk";
import { LocalStorage } from "node-localstorage";

export default class FieldStorage {
  public localStorage: LocalStorage;

  constructor(localStorage: LocalStorage) {
    this.localStorage = localStorage;
  }

  save(field: Field, id: string) {
    // saves a Field
    const compressedField = this.compress(field);
    // console.log(`saving field, ${compressedField}`)
    this.localStorage.setItem(id, compressedField);
    // console.log(`saved: ${compressedField}`);
  }
  load(id: string) {
    // returns a Field
    const compressedField = this.localStorage.getItem(id);
    if (compressedField === null)
      throw new Error(`Could not find field with id '${id}' in localStorage`);
    // console.log(compressedField)
    const field = this.decompress(compressedField, id);
    return field;
  }
  registerAutoSave(field: Field) {
    field.on("save", (chunks: any[]) => {
      console.log("saving field")
      //FieldStorage.save(field, saveName);
      chunks.forEach((chunk: any) => {
        this.saveChunk(chunk, field.fieldName);
      });
      this.save(field, field.fieldName);
    });
  }
  compress(field: Field) {
    // returns JSON string;
    const stringifiedField = JSON.stringify(field);
    // FieldStorage.logStats(field, stringifiedField);

    return stringifiedField;
  }
  saveChunk(chunk: Chunk, id: any) {
    this.localStorage.setItem(
      id + chunk.x + ";" + chunk.y,
      JSON.stringify(chunk)
    );
  }
  loadChunk(
    id: string,
    x: number,
    y: number,
    field?: Field | undefined
  ): Chunk | undefined {
    var chunk = new Chunk(x, y, field);
    if (this.localStorage.getItem(id + chunk.x + ";" + chunk.y)) {
      var data = JSON.parse(
        this.localStorage.getItem(id + chunk.x + ";" + chunk.y)
      );

      for (let i = 0; i < CHUNK_SIZE; i++) {
        for (let j = 0; j < CHUNK_SIZE; j++) {
          let cell = new Cell(x * CHUNK_SIZE + i, y * CHUNK_SIZE + j, field);

          let cellPointer = (i * CHUNK_SIZE + j) * 3;
          cell.isOpen = data.charAt(cellPointer) == true;
          const isMine = data.charAt(cellPointer + 1);
          cell.isMine =
            isMine == "2" ? undefined : isMine == "1" ? true : false;
          cell.isFlagged = data.charAt(cellPointer + 2) == true;
          chunk.cells[i][j] = cell;
        }
      }
      // console.log(`loaded chunk (${x},${y}): ${chunk}`);
      return chunk;
    }
    // console.log(`loading chunk (${x},${y}) is not set in localstorage, returning undefined`);
    return undefined;
  }
  decompress(compressedField: string, fieldName: string) {
    // when stringifying, we have changed the class into an object, and we
    // need to recreate the class from the data
    let recoveredField = JSON.parse(compressedField);
    let field = new Field(
      recoveredField.probability,
      recoveredField.safeRadius,
      this,
      fieldName
    );
    // Object.assign(field, recoveredField);
    field.score = recoveredField.score;
    if(recoveredField.pristine) throw new Error("trying to load pristine field. create a new instead")
    field.pristine = false;
    // field.probability = recoveredField.prhobability;
    return field;
  }
  logStats(
    field: { getAll: () => { (): any; new (): any; length: any } },
    string: string
  ) {
    const cellsCount = field.getAll().length;
    const compressedByteCount = unescape(encodeURI(string)).length;
    const ratio = compressedByteCount / cellsCount;
    console.log(
      `saved ${compressedByteCount} bytes with a compression ratio of ${ratio.toPrecision(
        5
      )} bytes/cell`
    );
    console.log(string);
  }
}
