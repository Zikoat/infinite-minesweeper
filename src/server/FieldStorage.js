import Field from "./Field.js";
import Cell from "./Cell.js";
import {Chunk} from "./Chunk.js";
import {CHUNK_SIZE} from "./Chunk.js";

export default class FieldStorage {
	static save(field, id) { // saves a Field
		const compressedField = FieldStorage.compress(field);
		localStorage.setItem(id, compressedField);
		// console.log(`saved: ${compressedField}`);
	}
	static load(id) { // returns a Field
		const compressedField = localStorage.getItem(id);
		const field = FieldStorage.decompress(compressedField);
		return field;
	}
	static registerAutoSave(field, saveName) {
		let timeSinceLastSave = 0;
		let lastInteractionTime = Date.now();
		FieldStorage.timerID = 0;

		field.on("cellChanged", ()=>{
			clearTimeout(FieldStorage.timerID);
			FieldStorage.timerID = setTimeout(()=>{
				FieldStorage.save(field, saveName);
			}, 1000);
		});

	}
	static compress(field) { // returns JSON string;
		const stringifiedField = JSON.stringify(field);
		// FieldStorage.logStats(field, stringifiedField);

		return stringifiedField;
	}
	static decompress(compressedField) {
		// when stringifying, we have changed the class into an object, and we 
		// need to recreate the class from the data
		let recoveredField = JSON.parse(compressedField);
		let field = new Field();
		
		for (const property in recoveredField) {
			field[property] = recoveredField[property];	
		}
		for (const row in recoveredField.field) {
			for (const column in recoveredField.field[row]) {
				const recoveredChunk = recoveredField.field[row][column];
				let chunk = new Chunk(parseInt(row), parseInt(column), field);
				for (const row1 in recoveredField.field[row][column].cells) {
					for (const column1 in recoveredField.field[row][column].cells[row1]) {
						const recoveredCell = recoveredField.field[row][column].cells[row1][column1];
						let cell = new Cell(parseInt(row1)+parseInt(row)*CHUNK_SIZE, parseInt(column1)+parseInt(column)*CHUNK_SIZE, field);
				

						cell.isOpen = (recoveredCell.charAt(0) == true);
						const isMine = recoveredCell.charAt(1);
						cell.isMine = isMine == "2"?undefined:isMine == "1"?true:false;
						cell.isFlagged = (recoveredCell.charAt(2) == true);
						chunk.cells[row1][column1] = cell;
					}
				}
				field.field[row][column] = chunk;
			}
		}

		return field;
	}
	static logStats(field, string) {
		const cellsCount = field.getAll().length;
		const compressedByteCount = unescape(encodeURI(string)).length;
		const ratio = compressedByteCount / cellsCount;
		console.log(`saved ${compressedByteCount} bytes with a compression ratio of ${ratio.toPrecision(5)} bytes/cell`);
		console.log(string);
	}
}