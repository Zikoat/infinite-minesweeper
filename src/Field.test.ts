// tests/demo.js
import { LocalStorage } from "node-localstorage";
import { suite, test } from "uvu";
import * as assert from "uvu/assert";
import Cell from "./Cell";
import Field, { ChunkedField } from "./Field";
import FieldStorage from "./FieldStorage";
import seedrandom from "seedrandom";
import { Chunk } from "./Chunk";

test("Math.sqrt()", () => {
  assert.is(Math.sqrt(4), 2);
  assert.is(Math.sqrt(144), 12);
  assert.is(Math.sqrt(2), Math.SQRT2);
});

test("JSON", () => {
  const input = {
    foo: "hello",
    bar: "world",
  };

  const output = JSON.stringify(input);

  assert.snapshot(output, `{"foo":"hello","bar":"world"}`);
  assert.equal(JSON.parse(output), input, "matches original");
});

test("Field should be able to be constructed", () => {
  const field = new Field(undefined, undefined, undefined, undefined);
  assert.is(field.gameOver, false);
});

test("Field should open a cell", () => {
  // todo remove. fieldstorage doesnt get cleaned up, and this test should be in the fieldstorage suite
  const fieldStorage = new FieldStorage(new LocalStorage("./localStorage"));
  const field = new Field(undefined, undefined, fieldStorage, "unitTest");
  const opened = field.open(0, 0);
  assert.is(opened, true);
});

test("Field should calculate score correctly", () => {
  const id = "unitTest";

  const fieldStorage = new FieldStorage(new LocalStorage("./localStorage"));
  const field = new Field(undefined, undefined, fieldStorage, id);
  assert.is(field.score, 0);
  const opened = field.open(0, 0);
  assert.is(opened, true);
  assert.is(field.score, 1);
});

test("Cell should be able to be instantiated", () => {
  const cell = new Cell(0, 0, false, false);
  assert.is(cell.isOpen, false);
});

const localStorage = new LocalStorage("./localstorage");
const localStorageSuite = suite("LocalStorage", localStorage);

localStorageSuite("should work like the example", (localStorage) => {
  localStorage.setItem("myFirstKey", "myFirstValue");

  assert.is(localStorage.getItem("myFirstKey"), "myFirstValue");

  const localStorage2 = new LocalStorage("./localstorage");
  assert.is(localStorage2.getItem("myFirstKey"), "myFirstValue");
});

localStorageSuite(
  "should return null for things that are not set",
  (localStorage) => {
    assert.is(localStorage.getItem("not defined"), null);
  }
);

localStorageSuite.after.each((localStorage) => {
  localStorage.clear();
  assert.is(localStorage.length, 0);
});

const fieldStorageSuite = suite(
  "FieldStorage",
  new FieldStorage(new LocalStorage("./localstorage"))
);
fieldStorageSuite(
  "should get an exact copy of the previous field",
  (fieldStorage) => {
    const field1 = new Field(0.6, 2, fieldStorage, "test1", "testSeed");

    assert.is(
      fieldStorage.localStorage.length,
      1,
      "localstorage length before open"
    );

    field1.open(0, 0);

    assert.is(
      fieldStorage.localStorage.length,
      1,
      "localstorage length after open"
    );
    assert.is(field1.fieldName, "test1");
    assert.is(field1.score, 9, "field score to static");
    assert.is(field1.getAll().length, 4096);
    assert.is(field1.gameOver, false);
    assert.is(field1.neighborPosition.length, 8);
    assert.is(field1.pristine, false);
    assert.is(field1.fieldStorage.localStorage.length, 1);

    field1.chunksToSave.slice(0).forEach((chunk: any) => {
      fieldStorage.saveChunk(chunk, field1.fieldName);
    });
    field1.chunksToSave = [];

    fieldStorage.save(field1, "test1");
    const field2 = fieldStorage.load("test1");

    // delete field1.fieldStorage;
    // delete field2.fieldStorage;
    // assert.equal(getAllKeys(field1.fieldStorage?.localStorage), ["test1"]);

    const cell1 = field1.getCell(0, 1);
    const cell2 = field2.getCell(0, 1);

    assert.is(cell1.isOpen, true, "cell1 open");
    assert.is(cell2.isOpen, true, "cell2 open");
    assert.is(field1.value(cell1.x, cell1.y), field2.value(cell2.x, cell2.y));

    fieldsAreEqual(field1, field2);
  }
);

export function getAllKeys(localStorage: LocalStorage): string[] {
  const keys: string[] = [];
  let i = 1;
  let key: boolean | string = true;
  while (key) {
    key = localStorage.key(i);

    if (key) keys.push(key);
    i++;
    if (i > 100) throw new Error("i is more than 100");
  }

  return keys;
}

function fieldsAreEqual(field1: Field, field2: Field): void {
  assert.equal(
    fieldViewToString(field2, -3, -3, 3, 3),
    `.......
.xxx...
..422x.
.x202..
.x324x.
...xxx.
.......
`
  );
  assert.equal(
    fieldViewToString(field1, -3, -3, 3, 3),
    fieldViewToString(field2, -3, -3, 3, 3)
  );

  assert.equal(
    getChunksInField(field2.fieldData),
    [
      [0, 0],
      [0, -1],
      [-1, 0],
      [-1, -1],
    ],
    "getchunksfield 1"
  );
  // assert.is(field1.field, field2.field)
  assert.is(field1.fieldName, field2.fieldName);
  assert.is(field1.score, field2.score, "field scores");
  assert.is(
    field1.getAll().length,
    field2.getAll().length,
    "fieldsareequal getall.length"
  );
  assert.is(field1.gameOver, field2.gameOver, "field gameover");
  assert.is(
    field1.neighborPosition,
    field2.neighborPosition,
    "field neighborpositions"
  );
  assert.is(field1.pristine, field2.pristine, "field pristine");
  assert.is(
    field1.fieldStorage.localStorage.length,
    field2.fieldStorage.localStorage.length
  );
}

function getChunksInField(chunkedField: ChunkedField): number[][] {
  const chunkCoords = [];
  for (const xChunks of Object.keys(chunkedField)) {
    for (const yChunks of Object.keys(chunkedField[parseInt(xChunks)])) {
      chunkCoords.push([parseInt(xChunks), parseInt(yChunks)]);
    }
  }
  return chunkCoords;
}

fieldStorageSuite.skip(
  "should not get call stack size exceeded when a loaded field opens a cell",
  (fieldStorage) => {
    const id = "unitTest";
    const field1 = new Field(undefined, 1, fieldStorage, id, "test");
    assert.is(field1.open(0, 0), true);
    field1.fieldStorage.__test__ = "";
    field1.fieldStorage.localStorage._eventUrl = "1";
    const expectedValue = {
      _events: {},
      _eventsCount: 0,
      probability: 0.5,
      pristine: false,
      safeRadius: 1,
      gameOver: false,
      neighborPosition: [
        [-1, 1],
        [0, 1],
        [1, 1],
        [-1, 0],
        [1, 0],
        [-1, -1],
        [0, -1],
        [1, -1],
      ],
      score: 1,
      chunksToSave: [
        "100000020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020010010020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020",
        "020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020010010020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020",
        "020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020000",
        "020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020010020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020000020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020020",
      ],
      visibleChunks: [],
      fieldStorage: {
        localStorage: {
          _location: "/home/zikoat/infinite-minesweeper/localstorage",
          quota: 5242880,
          _events: {},
          _eventsCount: 0,
          _maxListeners: null,
          length: 2,
          _bytesInUse: 12,
          _keys: ["__test__", "__suite__"],
          _metaKeyMap: {
            __test__: { key: "__test__", index: 0, size: 0 },
            __suite__: { key: "__suite__", index: 1, size: 12 },
          },
          _eventUrl: "1",
        },
        __test__: "",
        __suite__: "FieldStorage",
      },
      fieldName: "unitTest",
    };
    assert.snapshot(
      JSON.stringify(expectedValue, undefined, 2),
      JSON.stringify(field1, undefined, 2)
    );
    const scoreBefore = field1.score;
    fieldStorage.save(field1, id);
    const field2 = fieldStorage.load(id);

    fieldStorage.registerAutoSave(field1, id);
    const scoreAfter = field2.score;
    assert.is(scoreBefore, scoreAfter);
    const opened = field2.open(0, 0);
    assert.is(opened, true);
  }
);

fieldStorageSuite.after.each((fieldStorage) => {
  fieldStorage.localStorage.clear();
});

test("SeddedRandom should get the same random number every time", () => {
  const wow = seedrandom("shit");
  const randomNumber = wow();
  assert.is(randomNumber, 0.4459846419098659);
});

test("Chunk should be instantiated", () => {
  const chunk = new Chunk(0, 0);
  const chunkJsonFunction = chunk.toJSON();
  const chunkJsonStringified = JSON.stringify(chunk);
  assert.type(chunkJsonFunction, "string");
  assert.type(chunkJsonStringified, "string");
  assert.equal(
    chunkJsonFunction,
    chunkJsonStringified.substring(1, chunkJsonStringified.length - 1)
  );
  assert.instance(chunk, Chunk);
});

test("Chunk should get cell", () => {
  const chunk = new Chunk(0, 0);

  const cell: Cell = chunk.getCell(0, 0);
  assert.equal(cell.toJSON(), "020", "tojson cell");
  assert.instance(cell, Cell, "is instance");
  assert.equal(
    cellToObject(cell),
    {
      isFlagged: undefined,
      isOpen: false,
      parent: undefined,
      x: 0,
      y: 0,
      isMine: undefined,
    },
    "celltoobject"
  );
});

function cellToObject(cell: Cell): {
  isFlagged: boolean;
  isOpen: boolean;
  parent: Field | undefined;
  x: number;
  y: number;
  isMine: boolean | undefined;
} {
  const { isFlagged, isOpen, parent, x, y, isMine } = cell;
  return { isFlagged, isOpen, parent, x, y, isMine };
}

fieldStorageSuite("should be able to save and load a chunk", (fieldStorage) => {
  const chunk = new Chunk(0, 0);
  fieldStorage.saveChunk(chunk, "test");
  const loadedChunk = fieldStorage.loadChunk("test", 0, 0);
  assert.is(loadedChunk.getAll().length, 1024);

  for (const cell of loadedChunk.getAll()) {
    assert.is(cell.isFlagged, false);
    assert.is(cell.isMine, undefined);
    assert.is(cell.isOpen, false);
    assert.is(cell.parent, undefined);
    assert.ok(
      cell.x >= 0,
      JSON.stringify(cellToObject(cell)) + " does not have x>=0"
    );
    assert.ok(cell.x < 32, cellToObject(cell) + " does not have x<32");
    assert.ok(cell.y >= 0, cellToObject(cell) + " does not have y>=0");
    assert.ok(cell.y < 32, cellToObject(cell) + " does not have y<32");
  }
});

fieldStorageSuite("Field should draw a view", (fieldStorage) => {
  const field1 = new Field(0.6, 2, fieldStorage, "test1", "testSeed");
  field1.open(0, 0);

  const map = fieldViewToString(field1, -3, -3, 3, 3);

  assert.equal(
    map,
    `.......
.xxx...
..422x.
.x202..
.x324x.
...xxx.
.......
`
  );
});

function fieldViewToString(
  field: Field,
  x0: number,
  y0: number,
  x1: number,
  y1: number
): string {
  let map = "";

  for (let x = x0; x <= x1; x++) {
    for (let y = y0; y <= y1; y++) {
      const cell = field.getCell(x, y);
      let character = "";
      if (cell.isMine && cell.isOpen && !cell.isFlagged) character = "X";
      else if (cell.isMine && !cell.isOpen && cell.isFlagged) character = "F";
      else if (!cell.isMine && cell.isOpen && !cell.isFlagged) {
        const value = field.value(cell.x, cell.y);
        if (value === null) throw Error("the cell's value is null");
        character = value.toString();
      } else if (cell.isMine && !cell.isOpen && !cell.isFlagged)
        character = "x";
      else if (!cell.isMine && !cell.isOpen && !cell.isFlagged) character = ".";
      else
        throw new Error(
          `don't know what cell ${cell} on coordinate ${x},${y} is supposed to be`
        );

      map += character;
    }
    map += "\n";
  }
  return map;
}

test.run();
localStorageSuite.run();
fieldStorageSuite.run();
