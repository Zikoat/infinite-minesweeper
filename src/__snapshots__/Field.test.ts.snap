// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON: "{\"foo\":\"hello\",\"bar\":\"world\"}" 1`] = `"{"foo":"hello","bar":"world"}"`;

exports[`JSON 1`] = `"{"foo":"hello","bar":"world"}"`;
